R/allelic_series_sumstats.R

```diff
@@ -0,0 +1,331 @@
+  n_snp <- length(beta)
+  beta <- beta / skat_weights
+  se <- se / skat_weights
+  
+  # Calculate score test statistic from summary statistics.
+  w <- 1 / se
+  z <- beta / se
+  u <- w * z
+  
+  # Covariance of Z scores.
+  if (!is.null(ld)) {
+    cov_z <- (w %*% t(w)) * ld
+  } else{
+    cov_z <- diag(w^2)
+  }
+  
+  # SKAT-O grid search (provided by SKAT-O paper).
+  rhos <- c(0, 0.1^2, 0.2^2, 0.3^2, 0.5^2, 0.5, 0.999) 
+  n_rho <- length(rhos)
+  
+  # Q statistic: interpolates burden and variance comp statistics.
+  qstats <- (1 - rhos) * c(t(u) %*% (u)) + rhos * sum(u)^2
+  
+  # Loop over values of rho.
+  lambdas <- lapply(seq_len(n_rho), function(i) {
+    rho <- rhos[i]
+    cor_mat <- diag(rep(1 - rho, n_snp)) + matrix(rho, nrow = n_snp, ncol = n_snp)
+    cor_mat_sqrt <- chol(cor_mat, pivot = TRUE)
+    kernel <- cor_mat_sqrt %*% cov_z %*% t(cor_mat_sqrt)
+    out <- as.numeric(GetLambda(kernel))
+    return(out)
+  })
+
+  # Test parameters.
+  opt_params <- SkatOptimalParam(cov_z = cov_z, rhos = rhos)
+  results <- PerRhoResults(lambdas = lambdas, qstats = qstats, rhos = rhos)
+  pval <- OptimalPval(opt_params = opt_params, results = results)
+  return(pval)
+}
+
+
+# ------------------------------------------------------------------------------
+
+
+#' COding-variant Allelic Series Test from Summary Statistics
+#'
+#' Main function for performing the allelic series test from summary statistics.
+#' Performs both Burden and SKAT type tests, then combines the results to
+#' calculate an omnibus p-value. Note that not all tests included in
+#' \code{\link{COAST}} are available when working with summary statistics.
+#'
+#' @param anno (snps x 1) annotation vector with values in c(0, 1, 2).
+#' @param beta (snps x 1) vector of effect sizes for the coding genetic variants
+#'   within a gene.
+#' @param se (snps x 1) vector of standard errors for the effect sizes.
+#' @param check Run input checks? Default: TRUE.
+#' @param ld (snps x snps) matrix of correlations among the genetic variants.
+#'   Although ideally provided, an identity matrix is assumed if not.
+#' @param maf (snps x 1) vector of minor allele frequencies. Although ideally
+#'   provided, defaults to the zero vector.
+#' @param pval_weights (3 x 1) vector of relative weights for combining the
+#'   component tests when performing the omnibus test.
+#' @param weights (3 x 1) vector of annotation category weights.
+#' @return Numeric p-value.
+#' @examples 
+#' # Generate data.
+#' data <- DGP(n = 1e3)
+#' sumstats <- CalcSumstats(data = data)
+#' 
+#' # Run the Coding-variant Allelic Series Test from summary statistics.
+#' results <- COASTSS(
+#'   anno = sumstats$anno,
+#'   beta = sumstats$sumstats$beta, 
+#'   maf = sumstats$maf,
+#'   se = sumstats$sumstats$se,
+#'   ld = sumstats$ld
+#' )
+#' show(results)
+#' @export
+COASTSS <- function(
+    anno,
+    beta, 
+    se,
+    check = TRUE,
+    maf = NULL,
+    ld = NULL,
+    pval_weights = c(1, 1, 1),
```

Comment:
The COAST function calculates omni_weights inside the function. Perhaps for consistency, we should add omni_weights as the option? Also for consistency, do we want equal default weights?

Reply:
Great point, I've added the ability to set `pval_weights` to the original `COAST` function as well.